broker/httpd/api.go

```diff
@@ -7,26 +7,51 @@ import (
 	"io"
 	"net"
 	"net/http"
+	"sync"
 	"time"
 
 	"github.com/Symantec/Dominator/lib/log"
 	"github.com/Symantec/cloud-gate/broker"
 	"github.com/Symantec/cloud-gate/broker/configuration"
 	"github.com/Symantec/cloud-gate/broker/staticconfiguration"
+	"github.com/Symantec/cloud-gate/lib/constants"
+	"github.com/cviecco/go-simple-oidc-auth/authhandler"
+
+	"golang.org/x/net/context"
 )
 
 type HtmlWriter interface {
 	WriteHtml(writer io.Writer)
 }
 
+type AuthCookie struct {
+	Username  string
+	ExpiresAt time.Time
+}
+
 type Server struct {
 	brokers      map[string]broker.Broker
 	config       *configuration.Configuration
 	htmlWriters  []HtmlWriter
 	htmlTemplate *template.Template
+	logger       log.DebugLogger
+	cookieMutex  sync.Mutex
+	authCookie   map[string]AuthCookie
+	authSource   *authhandler.SimpleOIDCAuth
 	staticConfig *staticconfiguration.StaticConfiguration
+}
 
-	logger log.DebugLogger
+func (s *Server) performStateCleanup(secsBetweenCleanup int) {
+	for {
+		s.cookieMutex.Lock()
```

Comment:
It would be safer to have a function that does one iteration, with a defer to unlock, and then have this function call the inner function for every iteration. This ensures someone later doesn't add a continue/break which doesn't unlock the mutex.